File: Algebras/FixedMRegion/fmr/Face.cpp
/* 
 * This file is part of libfmr
 * 
 * File:   Face.cpp
 * Author: Florian Heinz <fh@sysv.de>
 * 
 * Created on September 9, 2016, 3:21 PM

//paragraph [1] Title: [{\Large \bf \begin {center}] [\end {center}}]
//[TOC] [\tableofcontents]

[1] Implementation of the class ~Face~

[TOC]

1 Overview

The class ~Face~ represents a polygon, which may contain one or more
holes. Several faces make up a ~Region~

2 Includes and definitions

*/

#include "fmr_Face.h"
#include <iostream>

using namespace fmr;

/*
3 Default constructor

*/
Face::Face() : lastPointValid(false) {
}

/*
4 Construct from RList

Construct the main cycle and the holes from an RList representation.
Format: (<main cycle><holes>[*])
Example:
( 
( ( 0 0 ) ( 100 0 ) ( 100 100 ) ) \_! main cycle !\_
( ( 5 5 ) (  95 5 ) (  95  95 ) ) \_! hole 1     !\_
)

*/
Face::Face(RList& l) : lastPointValid(false) {
    readRList(l[0]);
    for (int i = 1; i < l.size(); i++) {
        Face hole;
        hole.readRList(l[i]);
        holes.push_back(hole);
    }
}

/*
5 ~readRList~

Construct a single cycle from an RList representation.
This can either be the main cycle or a hole cycle.

*/
void Face::readRList(RList& l) {
    for (int i = 0; i < l.size(); i++) {
        double x = (double) l[i][0];
        double y = (double) l[i][1];
        addPoint(Point(x, y));
    }
    close();
}

/*
6 ~addSeg~

Adds the next line segment to this cycle.
 
*/
void Face::addSeg(Seg s) {
    segs.push_back(s);
}

/*
7 ~addPoint~

Creates and adds a new line segment from point ~p~ and
the previous point (if there is one) to this cycle.

*/
void Face::addPoint(Point p) {
    if (!lastPointValid) // There was no previous point
        lastPointValid = true;
    else {
        Seg s(lastPoint, p);
        addSeg(s);
    }
    lastPoint = p; // Remember the last point
}

/*
8 ~close~

Close the cycle by creating a line segment between the
endpoint of the last segment and the startpoint of
the first segment.

*/
void Face::close() {
    if (segs.size() >= 2) {
        Point last = segs[segs.size() - 1].f;
        Point first = segs[0].i;
        addSeg(Seg(last, first));
    }
}

/*
9 ~transform~

Transform this face according to the given transformation unit ~tu~ at
fraction ~frac~ in the time interval (i.e. 0.5 is the middle of the
time interval). This is done by transforming each line segment of this
cycle.

*/
Face Face::transform(TransformationUnit& tu, double frac) {
    Face f;

    // Transform the main cycle
    for (int i = 0; i < segs.size(); i++) {
        f.addSeg(segs[i].transform(tu, frac));
    }

    // Transform and add each hole
    for (int i = 0; i < holes.size(); i++) {
        f.holes.push_back(holes[i].transform(tu, frac));
    }

    return f;
}

/*
10 Point ~inside~

Tests, if a point is inside this face. This is exactly the case, when
it is inside the main cycle and outside of all hole cycles.
Uses the winding number algorithm.

*/
bool Face::inside(Point p) {
    int wn = 0;
    for (int i = 0; i < segs.size(); i++) {
        Seg& s = segs[i];
        if (s.i.y <= p.y) {
            if (s.f.y > p.y) {
                if (s.sign(p) > 0)
                    wn++;
            }
        } else {
            if (s.f.y <= p.y) {
                if (s.sign(p) < 0)
                    wn--;
            }
        }
    }
    bool inside = (wn != 0);

    if (!inside) // Point is outside the main cycle
        return false;

    for (int i = 0; i < holes.size(); i++) {
        Face& h = holes[i];
        if (h.inside(p))    // Point is inside a hole cycle,
            return false;   // hence outside the face
    }

    return true; // Point is truly inside the face
}

/*
11 ~area~

Calculates the area of this face. Holes are not taken into
account.

*/
double Face::area() {
    double ret = 0;
    
    for (int nrseg = 0; nrseg < segs.size(); nrseg++) {
        Seg s = segs[nrseg];
        ret += (s.i.x*s.f.y - s.f.x*s.i.y);
    }
    
    return ret/2.0;
}

/*
12 ~centroidParams~

Determines two parameters used to calculate the centroid of a region.

*/
std::pair<double, double> Face::centroidParams() {
    double xs = 0, ys = 0;
    
    for (int nrseg = 0; nrseg < segs.size(); nrseg++) {
        Seg seg = segs[nrseg];
        xs += (seg.i.x + seg.f.x) * (seg.i.x*seg.f.y - seg.f.x*seg.i.y);
        ys += (seg.i.y + seg.f.y) * (seg.i.x*seg.f.y - seg.f.x*seg.i.y);
    }
    
    return std::pair<double, double>(xs, ys);
}

/*
13 ~boundingBox~

Returns the bounding box of this face.

*/
BoundingBox Face::boundingBox() {
    BoundingBox bb;
    
    for (unsigned int nrseg = 0; nrseg < segs.size(); nrseg++) {
        Seg& seg = segs[nrseg];
        bb.update(seg.i);
        bb.update(seg.f);
    }
    
    return bb;
}

/*
14 ~ToString~

Returns a string representation of this face.

*/
std::string Face::ToString() {
    std::string ret = "( ";

    for (int i = 0; i < segs.size(); i++) {
        Seg s = segs[i];
        ret += s.i.ToString() + "\n";
    }
    ret += " )\n";

    return ret;
}

/*
15 ~toRList~

Returns an ~RList~ representation of this face.

*/
RList Face::toRList() {
    RList ret;
    
    RList &mf = ret.nest();
    for (int nrseg = 0; nrseg < segs.size(); nrseg++) {
        Seg s = segs[nrseg];
        mf.append(s.i.toRList());
    }
    for (int nrhole = 0; nrhole < holes.size(); nrhole++) {
        Face& hole = holes[nrhole];
        RList &hl = ret.nest();
        for (int nrseg = 0; nrseg < segs.size(); nrseg++) {
            Seg s = hole.segs[nrseg];
            hl.append(s.i.toRList());
        }
    }
    
    return ret;
}

File: Javagui/viewer/hoese/algebras/fixedmregion/Face.java
package viewer.hoese.algebras.fixedmregion;

import java.awt.Polygon;
import java.util.LinkedList;
import java.util.List;

/**
 * Class Face represents a single face of a region.
 * This is a polygon with optionally one or more holes.
 *
 * @author Florian Heinz <fh@sysv.de>
 */
public class Face {
    /** The last point set when constructing a face from points */
    private Point lastPoint;
    /** List of segments of this face */
    private List<Seg> segs = new LinkedList();
    /** List of holes in this face */
    private List<Face> holes = new LinkedList();
    
    /**
     * Add the next point when constructing a face.
     * 
     * @param point Next point
     */
    public void addPoint(Point point) {
        if (lastPoint != null) {
            segs.add(new Seg(lastPoint, point));
        }
        lastPoint = point;
    }

    /**
     * Close the cycle.
     * Creates the last segment from the last point to the start point.
     */
    public void close() {
        if (segs.size() < 2) {
            return;
        }
        Seg s = segs.get(0);
        segs.add(new Seg(lastPoint, s.s));
    }
    
    /**
     * Get the list of segments of this face.
     * 
     * @return List of segments
     */
    public List<Seg> getSegments() {
        return segs;
    }
    
    /**
     * Create a nested list of segments of the main cycle.
     * 
     * @return Nested List of segments
     */
    public NL segsToNL() {
        NL ret = new NL();
        for (Seg s : segs) {
            NL nl2 = ret.nest();
            nl2.addNr(s.s.x);
            nl2.addNr(s.s.y);
        }
        
        return ret;
    }
    
    /**
     * Create a Nested List representation of this face (cycle and holes)
     * 
     * @return Nested List representation
     */
    public NL toNL() {
        NL ret = new NL();
        ret.addNL(segsToNL());
        
        for (Face f : holes) {
            ret.addNL(f.segsToNL());
        }
        
        return ret;
    }
    
    /**
     * Create a single cycle from its Nested List representation
     * 
     * @param nl Nested List representation
     * @return A Face constructed from this cycle
     */
    public static Face singleFromNL(NL nl) {
        Face f = new Face();
        
        for (int i = 0; i < nl.size(); i++) {
            NL n = nl.get(i);
            Point p = new Point(n.get(0).getNr(), n.get(1).getNr());
            f.addPoint(p);
        }
        f.close();
        return f;
    }
    
    /**
     * Create a face from its Nested List representation (cycle and holes)
     * 
     * @param nl Nested List representation
     * @return A Face (cycle and optional holes)
     */
    public static Face fromNL(NL nl) {
        Face f = singleFromNL(nl.get(0));
        for (int i = 1; i < nl.size(); i++)
            f.addHole(singleFromNL(nl.get(i)));
        
        return f;
    }

    /**
     * Add a new hole to this face.
     * 
     * @param hole Hole to add
     */
    public void addHole (Face hole) {
        holes.add(hole);
    }
    /**
     * Get all holes of this face.
     * 
     * @return List of holes
     */
    public List<Face> getHoles () {
        return holes;
    }


    private Polygon p;
    public boolean contains (Point point) {
	    if (p == null) {
		    p = new Polygon();
		    for (Seg s : segs) {
			    p.addPoint((int)s.s.x, (int)s.s.y);
		    }
	    }
	    return p.contains(point.x, point.y);
    }

    public boolean contains (Face f) {
	    return contains (f.getFirstPoint());
    }

    public Point getFirstPoint() {
	    return segs.get(0).s;
    }



    /**
     * Calculate the projection of this face unter transformation trans at
     * instant currentTime.
     * 
     * @param trans The transformation
     * @param currentTime Instant to project
     * @return resulting face
     */
    public Face project (FMRegionTrans trans, double currentTime) {
        Face ret = new Face();
        for (Seg s : segs) {
            Seg n = new FMSeg(s, trans).project((long)currentTime);
            ret.segs.add(n);
        }
        for (Face hole : holes) {
            ret.addHole(hole.project(trans, currentTime));
        }
        
        return ret;
    }
}
